### DISCUSSION TITLE
Issue with Inversify and Sprotty version updates

### DISCUSSION THREAD

**Discussion by david-pa** [eclipse-glsp/glsp#985] [Q&A]
I'm currently working on a project that uses GLSP, and I'm attempting to update `inversify ^5.0.1 -> ^6.0.0` and `sprotty ^0.12.0 -> ^0.13.0`. Please keep in mind these updates are on seperate branches, but I think they may be related and so I am creating this post to discuss them both.

With sprotty, I'm getting a strange error like what's below with the following code snippet in a class that extends `AbstractUIExtension`:

```
protected setPosition(containerElement: HTMLElement): void {
    let x = 0;
    let y = 0;

    if (this.task) {
      const bounds = getAbsoluteClientBounds(this.task, this.domHelper, this.viewerOptions);
      x = bounds.x + 5;
      y = bounds.y + 5;
    }

    containerElement.style.left = `${x}px`;
    containerElement.style.top = `${y}px`;
    containerElement.style.width = '200px';
  }
```

> Argument of type 'import("/Users/davidpa/repo/.../node_modules/sprotty/lib/base/views/dom-helper").DOMHelper' is not assignable to parameter of type 'import("/Users/davidpa/repo/.../node_modules/@eclipse-glsp/client/node_modules/sprotty/lib/base/views/dom-helper").DOMHelper'.
> Types have separate declarations of a private property 'viewerOptions'.

With inversify, I'm getting the following inside `glsp/di.config.ts`:

```
...

import { Container, ContainerModule } from 'inversify';
import 'sprotty/css/edit-label.css';

...

const graphEditorModule = new ContainerModule((bind, unbind, isBound, rebind) => {
  rebind(TYPES.ILogger).to(ConsoleLogger).inSingletonScope();
  rebind(TYPES.LogLevel).toConstantValue(LogLevel.warn);
  bind(TYPES.ISnapper).to(GridSnapper);
  bind(TYPES.ICommandPaletteActionProvider).to(RevealNamedElementActionProvider);
  bind(TYPES.IContextMenuItemProvider).to(DeleteElementContextMenuItemProvider);
  const context = { bind, unbind, isBound, rebind };
  configureDefaultModelElements(context);
  bind(ExpandHandler).toSelf().inSingletonScope();
  configureActionHandler(context, CollapseExpandAction.KIND, ExpandHandler);
  configureActionHandler(context, SelectAction.KIND, InputButtonOperationHandler);
  configureActionHandler(context, WorkspaceUpdateAction.KIND, WorkspaceUpdateHandler);

...

export default function createContainer(widgetId: string): Container {
  const container = createClientContainer(
    graphEditorModule,
    directTaskEditor,
    EditLabelUIContainer,
    ButtonOverlayContainer,
    GraphEditorToolPaletteContainer
  );
  overrideViewerOptions(container, {
    baseDiv: widgetId,
    hiddenDiv: `${widgetId}_hidden`
  });
  return container;
}
```
Originally there was this issue (which went away after removing references to `rebind` and `unbind`):
> 
> Argument of type 'import("/Users/davidpa/.../node_modules/inversify/lib/container/container").Container' is not assignable to parameter of type 'import("/Users/davidpa/.../node_modules/sprotty/node_modules/inversify/lib/container/container").Container'.
> 
> Property '_appliedMiddleware' is missing in type 'import("/Users/davidpa/.../node_modules/inversify/lib/container/container").Container' but required in type 'import("/Users/davidpa/repo/.../node_modules/sprotty/node_modules/inversify/lib/container/container").Container'.

then there was:

> Argument of type '{ bind: interfaces.Bind; isBound: interfaces.Unbind; }' is not assignable to parameter of type '{ bind: Bind; isBound: IsBound; }'.
>   Types of property 'bind' are incompatible.
>     Type 'import("/Users/davidpa/repo/.../node_modules/inversify/lib/interfaces/interfaces").interfaces.Bind' is not assignable to type 'import("/Users/davidpa/repo/.../node_modules/sprotty/node_modules/inversify/lib/interfaces/interfaces").interfaces.Bind'.

and

> Argument of type '{ bind: interfaces.Bind; isBound: interfaces.Unbind; }' is not assignable to parameter of type 'ContainerContext'.
>   The types returned by 'isBound(...)' are incompatible between these types.
>     Type 'void' is not assignable to type 'boolean'.

which I feel may be requiring the Sprotty update to function, but I'm not sure.

Does anyone have any ideas on how to resolve these errors related to these packages?

**Comment by tortmayr** [comment]
Hi @david-pa ,
in general it's not that straight forward to simply update the `sprotty`/`inversify` versions to a newer version than provided by GLSP. You might end up with multiple different resolved versions  and type incompatibilities (which seems to be the root cause for your problems).

However, yarn has a built-in mechanism to force resolution of a one specific dependency version. You could try, and use that to update to a new sprotty version. Simply add a `resolutions` block to your root package.json like this:

```json
{

  "resolutions": {
    "**/sprotty": "0.13.0"
  }
}
```
This should work for sprotty because there are no known API breaks between the 0.12.0 and 0.13.0 version.
However, for inversify I would not recommend that. Updating to new major version most certainly will introduce API breaks and I'm not sure whether the application will still work with the version.

**Reply by david-pa** [reply]
We're actually using NPM right now, do you know if there's anything that can be done via NPM to help with Sprotty?

**Reply by tortmayr** [reply]
Hi, 
as far as I know the npm counter part to `resolutions` is `overrides`:
https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides
